protocols/dcutr/src/behaviour.rs

```diff
@@ -362,20 +377,21 @@ impl Candidates {
         }
     }
 
-    fn add(&mut self, mut address: Multiaddr) {
+    /// Format the address and push it into the LruCache if it is not relayed.
+    ///
+    /// Returns the provided address formatted if it was pushed.
+    /// Returns `None` otherwise.
+    fn add(&mut self, mut address: Multiaddr) -> Option<Multiaddr> {
         if is_relayed(&address) {
-            return;
+            return None;
         }
 
         if address.iter().last() != Some(Protocol::P2p(self.me)) {
             address.push(Protocol::P2p(self.me));
         }
 
-        self.inner.push(address, ());
-    }
-
-    fn iter(&self) -> impl Iterator<Item = &Multiaddr> {
-        self.inner.iter().map(|(a, _)| a)
+        self.inner.push(address.clone(), ());
+        Some(address)
```

Comment:
```suggestion
        match self.inner.push(address.clone(), ()) {
            Some((addr, ())) if addr == address => None,
            _ => Some(address)
        }
```

Only return `Some` if the address wasn't already in the cache?